### PR TITLE
ddl: fix the data and indexes inconsistency during ctc

### PR DESCRIPTION
**Pull request by AilinKid**
Signed-off-by: AilinKid <314806019@qq.com>

### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/25319

### What is changed and how it works?

fix the data and indexes inconsistency during the CTC middle stage.
The case is that:
```
1: during the middle stage of CTC.
2: insert one row and delete it (write-reorg will fulfill the index for this row, and delete action should remove this row too)
3: since the delete logic didn't get the right datum for the changing column (no casting transmission)
4: removing the already existed index added in the last statement won't take effect 
```

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

- ddl: fix the data and indexes inconsistency during ctc

### CODE REVIEW

**Review by zimulala — APPROVED** [review]
LGTM

File: table/tables/tables.go [hunk 1]
```diff
@@ -1044,7 +1044,16 @@ func (t *TableCommon) RemoveRecord(ctx sessionctx.Context, h kv.Handle, r []type
...
 	// The table has non-public column and this column is doing the operation of "modify/change column".
 	if len(t.Columns) > len(r) && t.Columns[len(r)].ChangeStateInfo != nil {
-		r = append(r, r[t.Columns[len(r)].ChangeStateInfo.DependencyColumnOffset])
+		// See https://github.com/pingcap/tidb/issues/25319
+		// The changing column datum derived from related column should be casted here.
+		// Otherwise, the existed changing indexes will not be deleted.
+		relatedColDatum := r[t.Columns[len(r)].ChangeStateInfo.DependencyColumnOffset]
+		value, err := table.CastValue(ctx, relatedColDatum, t.Columns[len(r)].ColumnInfo, false, false)
+		if err != nil {
+			logutil.BgLogger().Info("update record cast value failed", zap.Any("col", t.Columns[len(r)]),
```
Comment (tangenta) [inline table/tables/tables.go:1053]: Why not return an error if the cast failed?

Comment (AilinKid) [reply table/tables/tables.go:1053]: Aha, yes

File: table/tables/tables.go [hunk 2]
```diff
@@ -1044,7 +1044,17 @@ func (t *TableCommon) RemoveRecord(ctx sessionctx.Context, h kv.Handle, r []type
 
 	// The table has non-public column and this column is doing the operation of "modify/change column".
 	if len(t.Columns) > len(r) && t.Columns[len(r)].ChangeStateInfo != nil {
-		r = append(r, r[t.Columns[len(r)].ChangeStateInfo.DependencyColumnOffset])
+		// See https://github.com/pingcap/tidb/issues/25319
```
Comment (wjhuang2016) [inline table/tables/tables.go:1047]: I don't think we need to post the issue.

**Review by wjhuang2016 — APPROVED** [review]

**Pull request merged**